runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,320 @@
+        )
+      then
+        refresh(bufnr, args.data.client_id)
+      end
+    end,
+  })
+
+  for _, client in
+    ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+  do
+    refresh(bufnr, client.id)
+  end
+
+  return bufstate
+end
+
+---@param winid integer
+---@param ranges lsp.FoldingRange[]
+local function foldclose(winid, ranges)
+  vim._with({ win = winid }, function()
+    -- Ensuring folding the smallest ranges first is not explicitly supported,
+    -- but since LSP servers often return the largest ranges first,
+    -- we traverse the ranges in reverse order here.
+    --
+    -- If sorting needs to be implemented,
+    -- the most suitable algorithm here is
+    -- [Timesort](https://wikipedia.org/wiki/Timsort).
+    --
+    -- "Hanc marginis exiguitas non caperet."
+    for i = #ranges, 1, -1 do
+      local range = ranges[i]
+      local start_lnum = range.startLine + 1
+      local end_lnum = range.endLine + 1
+      -- Folding a single line is not supported.
+      if start_lnum ~= end_lnum then
+        vim.cmd(start_lnum .. ',' .. end_lnum .. 'foldclose')
+      end
+    end
+  end)
+end
+
+--- Close all {kind} of folds in the the window with {winid}.
+---
+--- To automatically fold imports when opening a file, you can use an autocmd:
+---
+--- ```lua
+--- vim.api.nvim_create_autocmd('LspNotify', {
+---   callback = function(args)
+---     if args.data.method == 'textDocument/didOpen' then
+---       vim.lsp.folding_range.foldclose('imports', vim.fn.bufwinid(args.buf))
+---     end
+---   end,
+--- })
+--- ```
+---
+---@param kind lsp.FoldingRangeKind "comment" | "imports" | "region"
+---@param winid? integer Defaults to the current window.
+function M.foldclose(kind, winid)
+  vim.validate('kind', kind, 'string')
+  vim.validate('winid', winid, 'number', true)
+
+  winid = winid or api.nvim_get_current_win()
+  local bufnr = api.nvim_win_get_buf(winid)
+  local bufstate = bufstates[bufnr]
+  if not bufstate then
+    return
+  end
+  ---@cast bufstate vim.lsp.folding_range.BufState
+
+  -- Defer `foldclose()` if the buffer is not up-to-date.
+  if bufstate.version ~= util.buf_versions[bufnr] then
+    for _, client in
+      ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+    do
+      local request_id = refresh(bufnr, client.id)
+      api.nvim_create_autocmd('LspRequest', {
+        buffer = bufnr,
+        group = augroup_refresh,
+        callback = function(args)
+          if args.data.request_id == request_id and args.data.request.type == 'complete' then
+            local ranges = bufstate.kind_ranges[kind] or {}
+            foldclose(winid, ranges)
+            return true
+          end
+        end,
+      })
+    end
+    return
+  end
+
+  local ranges = bufstate.kind_ranges[kind] or {}
+  foldclose(winid, ranges)
+end
+
+---@return string
+function M._foldtext()
+  local bufnr = api.nvim_get_current_buf()
```

Comment:
Is this really doing anything useful? It just seems to be returning the text of the top of the folded region?

Reply:
It is to support `collapsedText` in `lsp.foldingRange` (proposed). To be honest, I haven't found a server that supports it, but I have a friend who is trying to implement LSP for Verilog and said this feature is somewhat useful. Considering it is very simple, I added it.

Reply:
Let's wait until it's actually part of the spec then; things might change too much until then.

(It's better to leave something as a follow-up than to risk adding an API we have to do the deprecation dance around.)

Reply:
It has already become part of the current specification. In [3.17](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#version_3_17_0). It's just that the server I commonly use does not support this feature.

Reply:
The fact it doesn't provide highlighting already makes it behind other foldtexts, so I'd omit this for now.

Reply:
`collapsed_text` is often used for folding regions (delimited by `#region/#endregion`).

Example from TypeScript:
![image](https://github.com/user-attachments/assets/d4bf335f-2909-4ecc-99b3-1f9537cab3d4)

I think that we can support it, I doubt that the spec will change a lot.

Reply:
(I just went by the "(proposed)" in the second comment. If it's already part of the spec, then that's different of course.)

Reply:
> (I just went by the "(proposed)" in the second comment. If it's already part of the spec, then that's different of course.)

That's fair! I just wanted to give an example of how it is used :)